File: Year3/programmingParadigms/T1/convert.hs
{-# LANGUAGE OverloadedStrings, UnicodeSyntax #-}
import System.Environment
import ClassyPrelude (singleton)
import System.FilePath.Posix  (takeBaseName)
import Data.List
import Data.Char
import Data.Text as T (Text, pack, unpack, replace, concat)
import Text.Megaparsec
import Text.Megaparsec.Char
import Data.Void (Void)
import Data.List.Split
import System.Directory
import System.Exit
import Data.Typeable
import Data.Maybe
import Data.Either

{-
Program to replace markup text with html. Some functionality was written with the help of the official Megaparsec tutorial at: https://markkarpov.com/tutorial/megaparsec.html
-}

type Parser = Parsec Void Text  -- parser type to prevent having to type out Parsec Void Text all the time

exclusionList = ["^", "%", "&", "#", "?", "/140", "/141", "/142", "/143", "/2 ", "/20", "/27", "/3 ", "/8 ", "/18", "/15", "/45"]
exclusionList2 = ["^", "%", "&", "#", "?", "/140", "/141", "/142", "/143", "/2 ", "/20", "/27", "/3 ", "/8 ", "/18", "/15", "/45"]

removeTashkeel xs = [x | x <- xs, not (x `elem` ['ً'..'ْ'])]    -- function to remove tashkeel

headers = ["/1", "/7", "/30", "/80"]

{-function to detect h1 pattern
return: returns text wrapped in header tags while removing markup-}
h1 :: Parser Text
h1 = try $ do           -- try to backtrack and not fail completely if parser doesn't succeed
    _ <- (string "/1 ")   -- look for /1 in text
    _ <- many spaceChar   -- check for 0 or more spaces
    text <- manyTill (anySingleBut '\n') (lookAhead (string " /1"))   -- any character except newline and /1
    _ <- many spaceChar
    _ <- (string "/1 ")
    space   -- look for space character
    _ <- many (char ':')  -- zero or more of : character
    return (T.concat ["<h1>", (pack text), "</h1>"])  -- return text wrapped with <h1></h1>

{-function to detect h2 pattern
return: returns text wrapped in header tags while removing markup-}
h2 :: Parser Text
h2 = do 
    _ <- (string "/7 ") 
    _ <- many spaceChar
    text <- manyTill anySingle (lookAhead (string " /7"))   -- any character until  /7 without consuming  /7
    _ <- many spaceChar
    _ <- (string "/7 ")
    space
    _ <- many (char ':')
    return (T.concat ["<h2>", (pack text), "</h2>"])

{-function to detect h3 pattern
return: returns text wrapped in header tags while removing markup-}
h3 :: Parser Text
h3 = do 
    _ <- (string "/30")
    _ <- many spaceChar
    text <- manyTill anySingle (lookAhead (string "/30"))
    _ <- many spaceChar
    _ <- (string "/30")
    space
    _ <- many (char ':')
    return (T.concat ["<h3>", (pack text), "</h3>"])

{-function to detect h4 pattern
return: returns text wrapped in header tags while removing markup-}
h4 :: Parser Text
h4 = do 
    _ <- (string "/80")
    _ <- many spaceChar
    text <- manyTill anySingle (lookAhead (string " /80"))
    _ <- many spaceChar
    _ <- (string "/80")
    space
    _ <- many (char ':')
    return (T.concat ["<h4>", (pack text), "</h4>"])

{-function to detect p1 pattern
return: returns text wrapped in paragraph tags while removing markup-}
p1 :: Parser Text
p1 = do 
    at <- string "@"
    text <- (manyTill anySingle newline)
    return (T.concat ["<p>", at, (pack text), "</p>"])

{-function to remove tags in exclusionList2 pattern
return: returns empty string-}
exclusionListRemoval :: Parser Text
exclusionListRemoval = do 
              tag <- choice exclusionList2
              return ""

{-function to detect info1 pattern
return: returns text wrapped in span tags while removing markup-}
info1 :: Parser Text
info1 = try $ do
        _ <- (string "/13") <|> (string "/23")    -- parse /13 or /23
        text <- manyTill anySingle ((string "/13") <|> (string "/23"))  --parse until /13 or /23
        return (T.concat ["<span class=\"info1\" title=\"زيادة علم\">(&nbsp;", (pack text), "&nbsp;)</span>"])

{-function to detect info2 pattern
return: returns text wrapped in span tags while removing markup-}
info2 :: Parser Text
info2 = do
        tag <- (string "/10") <|> (string "/70") <|> (string "/72") <|> (string "/74") <|> (string "/75")
        text <- manyTill anySingle (string tag)
        return (T.concat ["<span class=\"info2\" title=\"زيادة علم\">(&nbsp;", (pack text), "&nbsp;)</span>"])

{-function to detect place pattern
return: returns text wrapped in span tags while removing markup-}
place :: Parser Text
place = do
        _ <- (string "/60")
        text <- manyTill anySingle (string "/60")
        return (T.concat ["<span class=\"place\" title=\"مكان\">", (pack text), "</span>"])

{-function to detect p2 pattern
return: returns text wrapped in span tags and adding <p> to the beginning while removing markup-}
p2 :: Parser Text
p2 = try $ do
    _ <- (char '$')
    paragraphNumber <- some numberChar    -- parse one or more numbers
    text <- (manyTill anySingle (lookAhead (Text.Megaparsec.oneOf ['ء'..'ٓ'])))    -- parse until character in range of ['ء'..'ٓ']
    haddathana <- manyTill (Text.Megaparsec.oneOf ['ء'..'ٓ']) (lookAhead (char ' '))    -- parse characters ['ء'..'ٓ'] until space
    return (T.concat ["<p><span class=\"number\">", (pack paragraphNumber), "</span>", "<span class=\"haddathana\">", pack haddathana, "</span>"])

{-function to detect p2End pattern
return: returns string to signify end of paragraph with </p>-}
p2End :: Parser Text
p2End = do
      _ <- char '*'
      return ".</p>"

{-function to detect author pattern
return: returns text wrapped in span tags while removing markup-}
author :: Parser Text
author = try $ do 
        _ <- string "/12"
        space1
        text <- manyTill anySingle (string "/12")
        return (T.concat ["<span class=\"author\" title=\"المصنِّف\">", (pack text), "</span>"])


specialList = ["/22", "/55", "/56", "/57", "/58", "/59", "/64", "/71", "/73"]   -- list of tags to look out for in special function

{-function to detect special pattern
return: returns text wrapped in span tags while removing markup-}
special :: Parser Text
special = try $ do
        tag <- choice specialList   -- parse one of tags in specialList
        space1  
        text <- manyTill anySingle (string tag)
        return (T.concat ["<span class=\"special\" title=\"عَلَم\">", pack text, "</span>"])

{-function to replace comma with ،
return: returns text form of ،-}
replaceComma :: Parser Text
replaceComma = do
              _ <- (char ',')
              return "،"

{-function to replace semicolon with ؛
return: returns text of ؛-}
replaceSemicolon :: Parser Text
replaceSemicolon = do
                _ <- (char ';')
                return "؛"

{-function to detect salah pattern
return: returns text of span wrapped condenced salah-}
replaceSalah :: Parser Text
replaceSalah = do
            _ <- (string "صلى الله عليه وسلم")
            return "<span class=\"salah\">ﷺ</span>"

{-function to detect quran mentions
return: returns text wrapped in span tags while removing markup-}
quranMention :: Parser Text
quranMention = do
        _ <- (string "/33") <|> (string "/65")
        text <- manyTill anySingle ((string "/33") <|> (string "/65"))
        return (T.concat ["<span class=\"quran\">", (pack text), "</span>"])

{-function to detect quran
return: returns text of quran wrapped in span tags and additional information while removing markup-}
quran :: Parser Text
quran = do
      _ <- (string "/4 ")
      first <- manyTill anySingle (lookAhead ("سورة"))
      stuff1 <- (string "سورة")
      second <- manyTill anySingle (lookAhead ("آية"))
      stuff2 <- (string "آية")
      third <- manyTill (numberChar <|> char '-' <|> spaceChar) (lookAhead (char '/'))
      _ <- (string "/4")
      return (T.concat ["<span class=\"quran\"><span class=\"leftmargin\">", (pack second), " ", (pack third), "</span>", (pack first), "</span>"])

{-function to detect quran without reference
return: returns text wrapped in span tags while removing markup-}
quranWithoutRef :: Parser Text
quranWithoutRef = do
                  _ <- (string "/4 ")
                  text <- manyTill (anySingle <|> spaceChar) (lookAhead ((string "/4")))
                  _ <- (string "/4") 
                  return (T.concat ["<span class=\"quran\">", (pack text), "</span>"])

{-function to detect narrator pattern
return: returns text wrapped in span tags with narrator information while removing markup-}
narrator :: Parser Text
narrator = try $ do
          tag <- (:) <$> char '/' <*> (some numberChar)   -- parse any /nr once
          if (tag) `elem` exclusionList then return "" else do  -- check if found tag is in exclusionList and return empty string if true
            text <- takeWhileP Nothing (\c -> c /= 'L')   -- take while character is not L
            _ <- char 'L'
            id <- (manyTill numberChar (lookAhead (char ' ')))
            _ <- manyTill spaceChar (lookAhead (char '/'))
            _ <- (string (pack tag))
            return (T.concat ["<span class=\"css_class_for_", (pack id), "_\" title=\"", (pack id), "\">", text, "</span>"])

inclusionList = ["/26", "/93", "/94"]

{-function to detect ungraded narrator pattern
return: returns text wrapped in span tags while removing markup-}
narratorUngraded :: Parser Text
narratorUngraded = try $ do
                  tag <- (:) <$> char '/' <*> (some numberChar)   -- parse any /nr once
                  if (tag) `elem` inclusionList then do   -- check if parsed tag is in inclusionList
                  text <- manyTill (anySingleBut 'L') (lookAhead (string (pack tag)))   -- if tag in inclusionList parse unless L until same tag is found again
                  _ <- (string (pack tag))
                  return (T.concat ["<span class=\"narrator_ungraded\" title=\"(راو)\">", (pack text), "</span>"])
                  else return ("")

{-function to find html tags denoting graded narrators and add narrator info taken from a csv file to the html
return: returns narrator text with added info in html tags-}
gradeNarrators :: [([Char], [[Char]])] -> Parser Text
gradeNarrators input = do
                _ <- (string "css_class_for_")
                id <- (manyTill numberChar (char '_'))
                let narrator = lookup id input    -- look for narrator in input data
                _ <- (string "\" title=\"")
                _ <- (manyTill numberChar (char '"')) 
                if narrator /= Nothing then do     -- if narrator is found in input
                    let name = (fromJust narrator)!!0   -- get narrator details and save to variables
                    let grade = (fromJust narrator)!!1
                    let gradeText = (fromJust narrator)!!2
                    let generation = (fromJust narrator)!!3
                    let total = (fromJust narrator)!!4
                    return (T.concat ["grade_color_", (pack grade), "\" title=\"", (pack name), "  -  ", (pack gradeText), "   (م", (pack grade), " - ط", (pack generation), " - ح", (pack total), ")\""])
                else do
                    return (T.concat ["not_in_db", " title=\"(", (pack id), "<\"(ليس في قاعدة البيانات "])  

{-function to detect poem verse 1 pattern
return: returns text wrapped in div tags while removing markup-}
poemVerse1 :: Parser Text
poemVerse1 = try $ do
            _ <- (string "/50")
            _ <- some spaceChar
            _ <- (string "/51")
            _ <- some spaceChar
            text <- manyTill (anySingle <|> spaceChar) (lookAhead (char '/'))
            _ <- (string "/51")
            _ <- some spaceChar
            _ <- (string "/50")
            _ <- many spaceChar
            _ <- many (char '،')
            _ <- many (char '.')
            return (T.concat ["<div class=\"poem_verse\">", (pack text), "</div>"])

{-function to detect poem verse 2 pattern
return: returns text wrapped in div tags while removing markup-}
poemVerse2 :: Parser Text
poemVerse2 = try $ do
            _ <- (string "/50")
            _ <- some spaceChar
            _ <- (string "/51")
            _ <- some spaceChar
            text1 <- manyTill (anySingle <|> spaceChar) (lookAhead (char '/'))
            _ <- (string "/51")
            _ <- some spaceChar
            _ <- (string "/51")
            text2 <- manyTill (anySingle <|> spaceChar) (lookAhead (char '/'))
            _ <- (string "/51")
            _ <- some spaceChar
            _ <- (string "/50")
            _ <- some spaceChar
            _ <- many (char '،')
            _ <- many (char '.')
            return (T.concat ["<div class=\"poem_verse\">", (pack text1), " ... ", (pack text2), "</div>"])

{-function to remove character E from text
return: returns empty text-}
eNumber :: Parser Text
eNumber = do
        _ <- (:) <$> char 'E' <*> (manyTill numberChar (lookAhead (char ' ')))
        return ""


{-function to detect haddathana pattern
return: returns text wrapped in span tags while removing markup-}
haddathana :: Parser Text
haddathana = try $ do
  _ <- lookAhead (string "@" <|> string"$")
  text <- lookAhead (manyTill anySingle (lookAhead (Text.Megaparsec.oneOf ['ء'..'ٓ'])))
  haddathana <- manyTill (Text.Megaparsec.oneOf ['ء'..'ٓ']) (lookAhead (char ' '))
  return $ (T.concat ["<span class=\"haddathana\">", pack haddathana, "</span>"])

{-function to detect page number pattern
return: returns text wrapped in span tags while removing markup-}
pageNumber :: Parser Text
pageNumber = try $ do
  _ <- char '@'
  space
  n <- some numberChar
  space
  _ <- char ':'
  space
  n2 <- some numberChar
  space
  _ <- char '@'
  return (T.concat ["<span class=\"pagenumber\">", (pack n), " / ", (pack n2), "</span>"])

{-function to parse text to remove markup and replace with html-}
parseText :: Parser [Text]
parseText = many $ 
            ((replaceSalah
            <|> replaceComma
            <|> replaceSemicolon
            <|> eNumber
            <|> h1
            <|> h2
            <|> h3
            <|> h4
            <|> p2End
            <|> p2
            <|> quran
            <|> quranWithoutRef
            <|> quranMention
            <|> special
            <|> place
            <|> poemVerse1
            <|> poemVerse2
            <|> narratorUngraded
            <|> narrator)  
            <|> (singleton <$> anySingle))    -- if none of the above parsers succeed or fit, parse single character

{-function to parse text to remove markup and replace with html. Separated to prevent parsing conflicts-}
parseSensitiveText :: Parser [Text]
parseSensitiveText = many $
  (pageNumber <|> haddathana <|> exclusionListRemoval <|> (singleton <$> anySingle))

{-function to parse text to remove markup and replace with html. Separated to prevent parsing conflicts-}
parseParagraph :: Parser [Text]
parseParagraph = many $ (p1 <|> (singleton <$> anySingle))

parseSensitiveText2 :: Parser [Text]
parseSensitiveText2 = many $ ((info1 <|> info2 <|> place <|> quran) <|> (singleton <$> anySingle))

{-function to concatenate parsed text back together from list-}
concatenateSensitiveText2 :: Parser Text
concatenateSensitiveText2 = T.concat <$> parseSensitiveText2

concatenateSensitiveText :: Parser Text
concatenateSensitiveText = T.concat <$> parseSensitiveText

{-function to parse and add narrator data to text-}
parseNarrators :: [([Char], [[Char]])] -> Parser [Text]
parseNarrators input = many $
  ((gradeNarrators input) <|> (singleton <$> anySingle))

concatenateText = T.concat <$> parseText

concatenateGradedText input = T.concat <$> (parseNarrators input)

concatenateParagraph = T.concat <$> parseParagraph

main = do
      file <- getArgs   -- get arguments provided by user
      if length file > 0 then   -- make sure user passed file name
            do
              boolFile <- doesFileExist ("input/" ++ file!!0)   -- make sure file exists
              if boolFile then
                    do
                    
                      template <- readFile "input/template.html"    -- get template information
                      contents <- readFile ("input/" ++ file!!0)    -- get text to convert from markup to html
                      narratorCSV <- readFile "input/narrators.csv"   -- get narrator information
                      
                      let narrators = map (\(x:xs) -> (x, xs)) (map (splitOn ",") (lines narratorCSV))    -- convert csv to a usable state
                      let initial = runParser (concatenateSensitiveText2) "file" (pack (removeTashkeel contents))   -- parse over text
                      
                      case initial of 
                        (Right initial) -> do   -- if parsing was successful
                          let result = runParser concatenateSensitiveText "file" initial    -- parse over text again
                          case result of
                            (Right result) -> do
                              let anotherResult = runParser concatenateText "file" result
                              case anotherResult of
                                (Right anotherResult) -> do
                                  let paragraphResult = runParser concatenateParagraph "file" anotherResult
                                  case paragraphResult of
                                    (Right paragraphResult) -> do
                                      let gradedResult = runParser (concatenateGradedText narrators) "file" paragraphResult   -- parse over text to add narrator grade details
                                      case gradedResult of
                                        (Right gradedResult) -> do
                                          let htmlTags = ["%%AUTHOR%%", "%%TITLE%%", "%%BODY%%"]    -- tags to replace with text or text information
                                          let changeTags (x:xs) = case x of { "%%TITLE%%" -> (replace x (pack (takeBaseName (file!!0))) (changeTags xs));
                                                                                        "%%AUTHOR%%" -> (replace x "" (changeTags xs));
                                                                                        "%%BODY%%" -> (replace x gradedResult (pack template)) }
                                          writeFile ("output/" ++ (takeBaseName (file!!0) ++ ".html")) (unpack (changeTags htmlTags))   -- write parsed text to new file in output folder
                                        Left err -> (putStrLn "Fail5")    -- show what parsing step parser failed at if parser fails
                                    Left err -> (putStrLn "Fail4")
                                Left err -> (putStrLn "Fail3")
                            Left err -> (putStrLn "Fail2")
                        Left err -> (putStrLn "Fail1")
              else
                exitWith (ExitFailure 1)
      else
        exitWith (ExitFailure 1)

File: Year3/programmingParadigms/T1/convert.py
import re, csv, sys

# -
# Read narrators data and store as dictionary for use later on
# -


def load_narrators():
    NARRATOR = dict()
    with open("input/narrators.csv", encoding='utf8') as file_narrators:
        csv_narrators = csv.DictReader(file_narrators)
        for row in csv_narrators:
            NARRATOR[row['ID']] = row
    return NARRATOR

# -
# Load the marked-up text, and start processing the mark-up using regular expressions
# -


re_h1 = re.compile(r'/1\s+(.+?)\s*/1(\s*:)?')
re_h2 = re.compile(r'/7\s+(.+?)\s*/7(\s*:)?')
re_h3 = re.compile(r'/30\s+(.+?)\s*/30(\s*:)?')
re_h4 = re.compile(r'/80\s+(.+?)\s*/80(\s*:)?')
re_p1 = re.compile(r'@\s*(.+?)$', re.MULTILINE)
re_p2 = re.compile(r'\$(\d+)[ 0-9/]+\s*(.+)\*')
re_narrator = re.compile(r'(/\d+)\s*([^/]+?)\s*L(\d+)\s*\1\b')
re_author = re.compile(r'/12\s+(.+?)\s+/12\b')
re_narrator_ungraded = re.compile(r'/(26|93|94)\s*([^/L]+?)\s*/\1\b')
re_special = re.compile(r'/(22|55|56|57|58|59|64|71|73)\s+(.+?)\s+/\1\b')
re_info1 = re.compile(r'/(13|23)\s+(.+?)\s+/\1\b')
re_info2 = re.compile(r'/(10|70|72|74|75)\s+(.+?)\s+/\1\b')
re_place = re.compile(r'/60\s+(.+?)\s+/60\b')
re_quran = re.compile(r'/4 (.+?) سورة (.+?) آية ([ 0-9-]+) /4\b') # use [^/]+? instead?
re_quran_without_ref = re.compile(r'/4 (.+?) /4\b') # use [^/]+? instead?
re_quran_mention = re.compile(r'/(33|65) (.+?) /\1\b')
re_page_number = re.compile(r'@\s*(\d+)\s*:\s*(\d+)\s*@')
re_poem_verse1 = re.compile(r'/50\s+/51\s+([^/]+)\s+/51\s+/50\s*[\.،]?')
re_poem_verse2 = re.compile(r'/50\s+/51\s+([^/]+)\s+/51\s+/51\s+([^/]+)\s+/51\s+/50\s*[\.،]?', re.DOTALL)
re_haddathana = re.compile(r'([\$@][^\u0621-\u0653]+?)([\u0621-\u0653]+) ')

re_remove_E_d = re.compile(r'E\d+')

# For Unicode see e.g. https://unicode-table.com/en/blocks/arabic/
re_arabic_without_tashkeel = re.compile(r'[^ \u0621-\u064a]')  # ء-ي  --  "tashkeel" -> short vowels
re_remove_tashkeel = re.compile(r'[\u064b-\u0652]')  # َ ً ُ ٌ ِ ٍ ْ ّ

def convert_markup_to_html(filename: str, NARRATOR: dict, remove_tashkeel=True):
    with open(f"input/{filename}", encoding='utf8') as textfile:
        text = textfile.read()

    if remove_tashkeel is True:
        text = re_remove_tashkeel.sub('', text)  # for simplicity...
        text = text.replace('صلى الله عليه وسلم', '<span class="salah">ﷺ</span>')
    text = text.replace(',', '،')
    text = text.replace(';', '؛')
    text = re_page_number.sub(r'<span class="pagenumber">\1 / \2</span>', text)
    text = re_haddathana.sub(r'\1<span class="haddathana">\2</span> ', text)
    # Ignored markup
    text = re_remove_E_d.sub('', text)
    for markup in ["^", "%", "&", "#", "/140 ", "/141 ", "/142 ", "/143", '/2 ', '/20', '/27', '/3 ', '/8 ', '/18 ', '/15 ', '/45 ']:
        text = text.replace(markup, '')

    # --
    text = re_h1.sub(r'<h1>\1</h1>', text)
    text = re_h2.sub(r'<h2>\1</h2>', text)
    text = re_h3.sub(r'<h3>\1</h3>', text)
    text = re_h4.sub(r'<b>\1</b>', text)  # todo
    text = re_p2.sub(r'<p><span class="number">\1</span>\2.</p>', text)
    text = re_p1.sub(r'<p>\1</p>', text)
    text = re_quran.sub(r'<span class="quran"><span class="leftmargin">\2 \3</span>\1</span>', text)
    text = re_quran_without_ref.sub(r'<span class="quran">\1</span>', text)
    text = re_quran_mention.sub(r'<span class="quran">\2</span>', text)
    text = re_special.sub(r'<span class="special" title="عَلَم">\2</span>', text)
    # Bug? nested tags...
    text = re_info1.sub(r'<span class="info1" title="زيادة علم">(&nbsp;\2&nbsp;)</span>', text)
    text = re_info2.sub(r'<span class="info2" title="زيادة علم">‹&nbsp;\2&nbsp;›</span>', text)
    text = re_author.sub(r'<span class="author" title="المصنِّف">\1</span>', text)
    text = re_place.sub(r'<span class="place" title="مكان">\1</span>', text)
    text = re_poem_verse1.sub(r'<div class="poem_verse">\1</div>', text)
    text = re_poem_verse2.sub(r'<div class="poem_verse">\1 ... \2</div>', text)
    # 'css_class_for_..._' will be transformed into a CSS class later on
    text = re_narrator.sub(r'<span class="css_class_for_\3_" title="\3">\2</span>', text)
    text = re_narrator_ungraded.sub(r'<span class="narrator_ungraded" title="(راو)">\2</span>', text)

    # -- Add narrators information

    re_css_for_narrator = re.compile(r'css_class_for_(\d+)_')
    unique_narrators_IDs = set(re_css_for_narrator.findall(text))
    for ID in unique_narrators_IDs:
        try:
            narrator = NARRATOR[ID]
            name = narrator['Name']
            grade = narrator['Grade']
            grade_text = narrator['Grade text']
            generation = narrator['Generation']
            total = narrator['Total']
        except KeyError:
            print(f"\tMissing narrator from database: ID={ID}")
            text = text.replace(f'css_class_for_{ID}_', 'not_in_db')
            text = text.replace(f'title="{ID}">', f'title="({ID} ليس في قاعدة البيانات)">')
            continue
        text = text.replace(f'css_class_for_{ID}_', f'grade_color_{grade}')
        text = text.replace(f'title="{ID}"', f'title="{name}  -  {grade_text}   (م{grade} - ط{generation} - ح{total})"') # طبقة - مرتبة - عدد أحاديثه

    # ---------------------------------

    with open("input/template.html", encoding='utf8') as file_template:
        HTML_TEMPLATE = file_template.read()
    HTML_TEMPLATE = HTML_TEMPLATE.replace('%%TITLE%%', re_arabic_without_tashkeel.sub('', filename))
    HTML_TEMPLATE = HTML_TEMPLATE.replace('%%AUTHOR%%', '')
    text = HTML_TEMPLATE.replace('%%BODY%%', text)
    with open(f"output/{filename[:-4]}.html", 'w', encoding='utf8') as file_html:
        file_html.write(text)


if __name__ == "__main__":
    try:
        filename = sys.argv[1]
        filename = re.sub(r'.+/([^/]+.txt)', r'\1', filename) # keep the file name only -- no path
    except IndexError:
        print("Please provide a filename.")
    
    NARRATOR = load_narrators()
    print("Converting:", filename)
    convert_markup_to_html(filename, NARRATOR)
